File: proposed-projects/round-trip-experiment/reverse/20260909T133702Z_8305aa82/20260910T075701Z_f285a71f/ported_implementation/src/rules-builder/rules-builder.ts
import { GrammarParseError } from '../utils/errors/grammar-parse-error.js';
import { isWordChar } from './is-word-char.js';
import { parseChar } from './parse-char.js';
import { parseName } from './parse-name.js';
import { parseSpace } from './parse-space.js';
import { SymbolIds } from './symbol-ids.js';
import { InternalRuleDef, InternalRuleType } from './types.js';

const WHITESPACE = /\s/;

export class RulesBuilder {
  public pos = 0;
  public symbolIds = new SymbolIds();
  // A sparse list: entries are `null` until the rule with that id is defined.
  public rules: (InternalRuleDef[] | null)[] = [];
  public src: string;
  public start: number;
  private timeLimit: number;

  constructor(src: string, limit = 1000) {
    this.src = src;
    this.start = performance.now();
    this.timeLimit = limit;
    this.parse(src);
  }

  private at(pos: number): string {
    return pos >= 0 && pos < this.src.length ? this.src[pos] : '';
  }

  parse(src: string): void {
    // move cursor forward until we reach non-whitespace content
    this.pos = parseSpace(src, 0, true);

    while (this.pos < src.length) {
      this.parseRule(src);
    }

    // Validate the state to ensure that all rules are defined
    for (const rule of this.rules) {
      if (rule === null) {
        continue;
      }
      for (const elem of rule) {
        if (elem.type === InternalRuleType.RULE_REF) {
          // Ensure that the rule at that location exists
          const value = elem.value as number;
          const referenced = value < this.rules.length ? this.rules[value] : null;
          const ruleExists = referenced !== null && referenced.length > 0;
          if (!ruleExists) {
            const missingRuleName = this.symbolIds.reverseGet(value);
            let missingRulePos = this.symbolIds.getPos(missingRuleName);

            // Skip over the ::= and any whitespace
            while (
              missingRulePos < src.length &&
              (src[missingRulePos] === ':' ||
                src[missingRulePos] === '=' ||
                WHITESPACE.test(src[missingRulePos]))
            ) {
              missingRulePos += 1;
            }

            throw new GrammarParseError(
              src,
              missingRulePos,
              `Undefined rule identifier "${missingRuleName}"`,
            );
          }
        }
      }
    }
  }

  parseRule(src: string): void {
    const name = parseName(src, this.pos);
    this.pos = parseSpace(src, this.pos + name.length, false);
    const ruleId = this.getSymbolId(name, name.length);

    // Skip over whitespace characters and find the ::= sequence
    this.pos = parseSpace(src, this.pos, true);
    if (
      !(
        this.at(this.pos) === ':' &&
        this.at(this.pos + 1) === ':' &&
        this.at(this.pos + 2) === '='
      )
    ) {
      throw new GrammarParseError(src, this.pos, `Expecting ::= at ${this.pos}`);
    }
    this.pos = parseSpace(src, this.pos + 3, true);

    this.parseAlternates(name, ruleId);

    if (this.at(this.pos) === '\r') {
      this.pos += this.at(this.pos + 1) === '\n' ? 2 : 1;
    } else if (this.at(this.pos) === '\n') {
      this.pos += 1;
    } else if (this.at(this.pos)) {
      throw new GrammarParseError(
        src,
        this.pos,
        `Expecting newline or end at ${this.pos}`,
      );
    }
    this.pos = parseSpace(src, this.pos, true);
  }

  getSymbolId(src: string, length: number): number {
    const nextId = this.symbolIds.size;
    const key = src.slice(0, length);
    if (!this.symbolIds.has(key)) {
      this.symbolIds.set(key, nextId, this.pos);
    }
    return this.symbolIds.get(key);
  }

  generateSymbolId(baseName: string): number {
    const nextId = this.symbolIds.size;
    this.symbolIds.set(`${baseName}_${nextId}`, nextId, this.pos);
    return nextId;
  }

  addRule(ruleId: number, rule: InternalRuleDef[]): void {
    while (this.rules.length <= ruleId) {
      this.rules.push(null);
    }
    this.rules[ruleId] = rule;
  }

  checkDuration(): void {
    if (performance.now() - this.start > this.timeLimit) {
      throw new GrammarParseError(
        this.src,
        this.pos,
        `duration of ${this.timeLimit} exceeded:`,
      );
    }
  }

  parseSequence(
    ruleName: string,
    outElements: InternalRuleDef[],
    depth = 0,
  ): void {
    const isNested = depth !== 0;
    const src = this.src;
    let lastSymStart = outElements.length;
    while (this.at(this.pos)) {
      const char = this.at(this.pos);
      if (char === '"') {
        this.pos += 1;
        lastSymStart = outElements.length;
        while (this.at(this.pos) !== '"') {
          this.checkDuration();
          const [value, incPos] = parseChar(src, this.pos);
          outElements.push(new InternalRuleDef(InternalRuleType.CHAR, [value]));
          // Adjusting pos by the length of parsed characters
          this.pos += incPos;
        }
        this.pos = parseSpace(src, this.pos + 1, isNested);
      } else if (char === '[') {
        this.pos += 1;
        let startType = InternalRuleType.CHAR;
        if (this.at(this.pos) === '^') {
          this.pos += 1;
          startType = InternalRuleType.CHAR_NOT;
        }
        lastSymStart = outElements.length;
        while (this.at(this.pos) !== ']') {
          this.checkDuration();
          const type =
            lastSymStart < outElements.length
              ? InternalRuleType.CHAR_ALT
              : startType;
          const [startcharValue, incPos] = parseChar(src, this.pos);
          this.pos += incPos;
          if (
            type === InternalRuleType.CHAR ||
            type === InternalRuleType.CHAR_NOT
          ) {
            outElements.push(new InternalRuleDef(type, [startcharValue]));
          } else {
            outElements.push(new InternalRuleDef(type, startcharValue));
          }

          if (this.at(this.pos) === '-' && this.at(this.pos + 1) !== ']') {
            this.pos += 1;
            const [endcharValue, endIncPos] = parseChar(src, this.pos);
            outElements.push(
              new InternalRuleDef(
                InternalRuleType.CHAR_RNG_UPPER,
                endcharValue,
              ),
            );
            this.pos += endIncPos;
          }
        }
        this.pos = parseSpace(src, this.pos + 1, isNested);
      } else if (isWordChar(char)) {
        const name = parseName(src, this.pos);
        const refRuleId = this.getSymbolId(name, name.length);
        this.pos += name.length;
        this.pos = parseSpace(src, this.pos, isNested);

        lastSymStart = outElements.length;
        outElements.push(
          new InternalRuleDef(InternalRuleType.RULE_REF, refRuleId),
        );
      } else if (char === '(') {
        this.pos = parseSpace(src, this.pos + 1, true);
        const subRuleId = this.generateSymbolId(ruleName);
        this.parseAlternates(ruleName, subRuleId, depth + 1);
        lastSymStart = outElements.length;
        outElements.push(
          new InternalRuleDef(InternalRuleType.RULE_REF, subRuleId),
        );
        if (this.at(this.pos) !== ')') {
          throw new GrammarParseError(
            src,
            this.pos,
            `Expecting ')' at ${this.pos}`,
          );
        }
        this.pos = parseSpace(src, this.pos + 1, isNested);
      } else if (char === '*' || char === '+' || char === '?') {
        if (lastSymStart === outElements.length) {
          throw new GrammarParseError(
            src,
            this.pos,
            `Expecting preceding item to */+/? at ${this.pos}`,
          );
        }
        const subRuleId = this.generateSymbolId(ruleName);
        const subRule = outElements.slice(lastSymStart);
        if (char === '*' || char === '+') {
          subRule.push(
            new InternalRuleDef(InternalRuleType.RULE_REF, subRuleId),
          );
        }
        subRule.push(new InternalRuleDef(InternalRuleType.ALT));
        if (char === '+') {
          subRule.push(...outElements.slice(lastSymStart));
        }
        subRule.push(new InternalRuleDef(InternalRuleType.END));
        this.addRule(subRuleId, subRule);
        outElements.length = lastSymStart;
        outElements.push(
          new InternalRuleDef(InternalRuleType.RULE_REF, subRuleId),
        );
        this.pos = parseSpace(src, this.pos + 1, isNested);
      } else {
        break;
      }
    }
  }

  parseAlternates(ruleName: string, ruleId: number, depth = 0): void {
    const src = this.src;
    const rule: InternalRuleDef[] = [];
    this.parseSequence(ruleName, rule, depth);
    while (this.at(this.pos) === '|') {
      this.checkDuration();
      rule.push(new InternalRuleDef(InternalRuleType.ALT));
      this.pos = parseSpace(src, this.pos + 1, true);
      this.parseSequence(ruleName, rule, depth);
    }
    rule.push(new InternalRuleDef(InternalRuleType.END));
    this.addRule(ruleId, rule);
  }
}

File: packages/gbnf-experiment/data/20260909T193332Z_f85fdfbb/ported_implementation/rules_builder/rules_builder.py
from __future__ import annotations

import re
import time

from ..utils.errors.grammar_parse_error import GrammarParseError
from ..utils.utf16 import to_utf16_units
from .is_word_char import is_word_char
from .parse_char import parse_char
from .parse_name import parse_name
from .parse_space import parse_space
from .symbol_ids import SymbolIds
from .types import (
    InternalRuleDef,
    InternalRuleDefAlt,
    InternalRuleDefChar,
    InternalRuleDefCharAlt,
    InternalRuleDefCharNot,
    InternalRuleDefCharRngUpper,
    InternalRuleDefEnd,
    InternalRuleDefReference,
    InternalRuleType,
)

_WHITESPACE = re.compile(r"\s")


def _now_ms() -> float:
    return time.perf_counter() * 1000


class RulesBuilder:
    """Parses a GBNF grammar into a flat list of rule definitions per symbol."""

    def __init__(self, src: str, limit: float = 1000):
        self.pos = 0
        self.symbol_ids: SymbolIds = SymbolIds()
        # A sparse list, mirroring the reference's sparse array: entries that were
        # referenced but never defined stay `None`.
        self.rules: list[list[InternalRuleDef] | None] = []
        self.src = to_utf16_units(src)
        self.start = _now_ms()
        self._time_limit = limit
        self._parse(self.src)

    # ------------------------------------------------------------------ parsing

    def _at(self, pos: int) -> str:
        src = self.src
        return src[pos] if 0 <= pos < len(src) else ""

    def _parse(self, src: str) -> None:
        # move cursor forward until we reach non-whitespace content
        self.pos = parse_space(src, 0, True)

        while self.pos < len(src):
            self._parse_rule(src)

        # Validate the state to ensure that all rules are defined
        for rule in self.rules:
            if rule is None:
                continue
            for elem in rule:
                if elem.type == InternalRuleType.RULE_REF:
                    # Ensure that the rule at that location exists
                    referenced = (
                        self.rules[elem.value] if elem.value < len(self.rules) else None
                    )
                    rule_exists = referenced is not None and len(referenced) > 0
                    if not rule_exists:
                        missing_rule_name = self.symbol_ids.reverse_get(elem.value)
                        missing_rule_pos = self.symbol_ids.get_pos(missing_rule_name)

                        # Skip over the ::= and any whitespace
                        while missing_rule_pos < len(src) and (
                            src[missing_rule_pos] in (":", "=")
                            or _WHITESPACE.search(src[missing_rule_pos])
                        ):
                            missing_rule_pos += 1

                        raise GrammarParseError(
                            src,
                            missing_rule_pos,
                            f'Undefined rule identifier "{missing_rule_name}"',
                        )

    def _parse_rule(self, src: str) -> None:
        name = parse_name(src, self.pos)
        self.pos = parse_space(src, self.pos + len(name), False)
        rule_id = self.get_symbol_id(name, len(name))

        # Skip over whitespace characters and find the ::= sequence
        self.pos = parse_space(src, self.pos, True)
        if not (
            self._at(self.pos) == ":"
            and self._at(self.pos + 1) == ":"
            and self._at(self.pos + 2) == "="
        ):
            raise GrammarParseError(src, self.pos, f"Expecting ::= at {self.pos}")
        self.pos = parse_space(src, self.pos + 3, True)

        self.parse_alternates(name, rule_id)

        if self._at(self.pos) == "\r":
            self.pos += 2 if self._at(self.pos + 1) == "\n" else 1
        elif self._at(self.pos) == "\n":
            self.pos += 1
        elif self._at(self.pos):
            raise GrammarParseError(
                src, self.pos, f"Expecting newline or end at {self.pos}"
            )
        self.pos = parse_space(src, self.pos, True)

    # ------------------------------------------------------------------- symbols

    def get_symbol_id(self, src: str, length: int) -> int:
        next_id = self.symbol_ids.size
        key = src[:length]
        if not self.symbol_ids.has(key):
            self.symbol_ids.set(key, next_id, self.pos)
        return self.symbol_ids.get(key)

    def generate_symbol_id(self, base_name: str) -> int:
        next_id = self.symbol_ids.size
        self.symbol_ids.set(f"{base_name}_{next_id}", next_id, self.pos)
        return next_id

    def add_rule(self, rule_id: int, rule: list[InternalRuleDef]) -> None:
        if rule_id >= len(self.rules):
            self.rules.extend([None] * (rule_id + 1 - len(self.rules)))
        self.rules[rule_id] = rule

    def check_duration(self) -> None:
        if _now_ms() - self.start > self._time_limit:
            raise GrammarParseError(
                self.src, self.pos, f"duration of {self._time_limit} exceeded:"
            )

    # ----------------------------------------------------------------- sequences

    def parse_sequence(
        self,
        rule_name: str,
        out_elements: list[InternalRuleDef],
        depth: int = 0,
    ) -> None:
        is_nested = depth != 0
        src = self.src
        last_sym_start = len(out_elements)
        while self._at(self.pos):
            char = self._at(self.pos)
            if char == '"':
                self.pos += 1
                last_sym_start = len(out_elements)
                while self._at(self.pos) != '"':
                    self.check_duration()
                    value, inc_pos = parse_char(src, self.pos)
                    out_elements.append(InternalRuleDefChar([value]))
                    # Adjusting pos by the length of parsed characters
                    self.pos += inc_pos
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif char == "[":
                self.pos += 1
                start_type = InternalRuleType.CHAR
                if self._at(self.pos) == "^":
                    self.pos += 1
                    start_type = InternalRuleType.CHAR_NOT
                last_sym_start = len(out_elements)
                while self._at(self.pos) != "]":
                    self.check_duration()
                    type_ = (
                        InternalRuleType.CHAR_ALT
                        if last_sym_start < len(out_elements)
                        else start_type
                    )
                    startchar_value, inc_pos = parse_char(src, self.pos)
                    self.pos += inc_pos
                    if type_ == InternalRuleType.CHAR:
                        out_elements.append(InternalRuleDefChar([startchar_value]))
                    elif type_ == InternalRuleType.CHAR_NOT:
                        out_elements.append(InternalRuleDefCharNot([startchar_value]))
                    else:
                        out_elements.append(InternalRuleDefCharAlt(startchar_value))

                    if self._at(self.pos) == "-" and self._at(self.pos + 1) != "]":
                        self.pos += 1
                        endchar_value, inc_pos = parse_char(src, self.pos)
                        out_elements.append(
                            InternalRuleDefCharRngUpper(endchar_value)
                        )
                        self.pos += inc_pos
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif is_word_char(char):
                name = parse_name(src, self.pos)
                ref_rule_id = self.get_symbol_id(name, len(name))
                self.pos += len(name)
                self.pos = parse_space(src, self.pos, is_nested)

                last_sym_start = len(out_elements)
                out_elements.append(InternalRuleDefReference(ref_rule_id))
            elif char == "(":
                self.pos = parse_space(src, self.pos + 1, True)
                sub_rule_id = self.generate_symbol_id(rule_name)
                self.parse_alternates(rule_name, sub_rule_id, depth + 1)
                last_sym_start = len(out_elements)
                out_elements.append(InternalRuleDefReference(sub_rule_id))
                if self._at(self.pos) != ")":
                    raise GrammarParseError(src, self.pos, f"Expecting ')' at {self.pos}")
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif char in ("*", "+", "?"):
                if last_sym_start == len(out_elements):
                    raise GrammarParseError(
                        src, self.pos, f"Expecting preceding item to */+/? at {self.pos}"
                    )
                sub_rule_id = self.generate_symbol_id(rule_name)
                sub_rule: list[InternalRuleDef] = out_elements[last_sym_start:]
                if char in ("*", "+"):
                    sub_rule.append(InternalRuleDefReference(sub_rule_id))
                sub_rule.append(InternalRuleDefAlt())
                if char == "+":
                    sub_rule.extend(out_elements[last_sym_start:])
                sub_rule.append(InternalRuleDefEnd())
                self.add_rule(sub_rule_id, sub_rule)
                del out_elements[last_sym_start:]
                out_elements.append(InternalRuleDefReference(sub_rule_id))
                self.pos = parse_space(src, self.pos + 1, is_nested)
            else:
                break

    def parse_alternates(self, rule_name: str, rule_id: int, depth: int = 0) -> None:
        src = self.src
        rule: list[InternalRuleDef] = []
        self.parse_sequence(rule_name, rule, depth)
        while self._at(self.pos) == "|":
            self.check_duration()
            rule.append(InternalRuleDefAlt())
            self.pos = parse_space(src, self.pos + 1, True)
            self.parse_sequence(rule_name, rule, depth)
        rule.append(InternalRuleDefEnd())
        self.add_rule(rule_id, rule)

    # camelCase aliases, mirroring the reference API.
    symbolIds = property(lambda self: self.symbol_ids)
    getSymbolId = get_symbol_id
    generateSymbolId = generate_symbol_id
    addRule = add_rule
    checkDuration = check_duration
    parseSequence = parse_sequence
    parseAlternates = parse_alternates
